### PR TITLE
Match AxiomIdentifier based on external syntax

### PR DESCRIPTION
**Pull request by ttuegel**
When `TermLike` patterns are `internalized`, we would not match `AxiomIdentifier`s for internal representations of Map, List, or Set. With this change, we always return an `AxiomIdentifier` based on the external pattern syntax, i.e. ordinary symbols.

Fixes: #1022 

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

**Comment by ttuegel** [comment]
This pull request is larger than I wanted because I had to rearrange some modules to avoid cycles (separating builtin simplifiers from the functions for manipulating their structure). I will make another pull request with all the refactoring separately.

File: kore/src/Kore/Step/Axiom/Identifier.hs [hunk 1]
```diff
@@ -85,24 +91,25 @@ Returns 'Nothing' if the 'TermLike' does not conform to one of the structures we
 recognize.
 
  -}
-matchAxiomIdentifier :: TermLike variable -> Maybe AxiomIdentifier
-matchAxiomIdentifier = Recursive.fold matchWorker
+matchAxiomIdentifier
+    :: InternalVariable variable
+    => TermLike variable
+    -> Maybe AxiomIdentifier
+matchAxiomIdentifier = Recursive.fold matchWorker . Builtin.externalize
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Axiom/Identifier.hs]: Is `externalize` evaluated lazily?

Comment (ttuegel) [reply kore/src/Kore/Step/Axiom/Identifier.hs]: Yes, it is defined by `Recursive.unfold`, so it lazily unfolds the `TermLike`.

**Comment by ttuegel** [comment]
Needs #1025.

**Pull request merged**